File: Provenance/Controller/iCade/PViCadeGamepadDirectionPad.m
//
//  PViCadeGamepadDirectionPad.m
//  Provenance
//
//  Created by Josejulio Martínez on 19/06/15.
//  Copyright (c) 2015 Josejulio Martínez. All rights reserved.
//

#import "PViCadeGamepadDirectionPad.h"
#import "PViCadeReader.h"

@interface PViCadeAxisInput(PVPrivateAPI)
-(void) setValue:(float) value;
@end

@interface PViCadeGamepadButtonInput(PVPrivateAPI)
-(void) setPressed:(BOOL)pressed;
@end

@implementation PViCadeGamepadDirectionPad

-(instancetype) init {
    if (self = [super init]) {
        _xAxis = [[PViCadeAxisInput alloc] init];
        _yAxis = [[PViCadeAxisInput alloc] init];

        _up = [[PViCadeGamepadButtonInput alloc] init];
        _down = [[PViCadeGamepadButtonInput alloc] init];
        _left = [[PViCadeGamepadButtonInput alloc] init];
        _right = [[PViCadeGamepadButtonInput alloc] init];
    }
    return self;
}

-(void) setValueChangedHandler:(GCControllerDirectionPadValueChangedHandler) handler {
    _handler = handler;
}

-(void) padChanged {
    iCadeState state = [PViCadeReader sharedReader].state;
    
    float x = (state & iCadeJoystickLeft)? -1.0f: ((state & iCadeJoystickRight)? 1.0f : 0.0f );
    float y = (state & iCadeJoystickDown)? -1.0f: ((state & iCadeJoystickUp)? 1.0f : 0.0f );
    [_xAxis setValue:x];
    [_yAxis setValue:y];

    if (x == -1.0f) {
        [_left setPressed:YES];
    } else if (x == 1.0f) {
        [_right setPressed:YES];
    } else {
        [_left setPressed:NO];
        [_right setPressed:NO];
    }

    if (y == -1.0f) {
        [_down setPressed:YES];
    } else if (y == 1.0f) {
        [_up setPressed:YES];
    } else {
        [_down setPressed:NO];
        [_up setPressed:NO];
    }

    if (_handler) {
        _handler(self, x, y);
    }
}

-(PViCadeAxisInput*) xAxis {
    return _xAxis;
}

-(PViCadeAxisInput*) yAxis {
    return _yAxis;
}

-(PViCadeGamepadButtonInput*) up
{
    return _up;
}

-(PViCadeGamepadButtonInput*) down;
{
    return _down;
}

-(PViCadeGamepadButtonInput*) left;
{
    return _left;
}

-(PViCadeGamepadButtonInput*) right;
{
    return _right;
}

@end

File: PVSupport/Sources/PVSupport/Controller/iCade/PViCadeGamepadDirectionPad.swift
//  Converted to Swift 4 by Swiftify v4.2.29618 - https://objectivec2swift.com/
//
//  PViCadeGamepadDirectionPad.swift
//  Provenance
//
//  Created by Josejulio Martínez on 19/06/15.
//  Copyright (c) 2015 Josejulio Martínez. All rights reserved.
//

#if canImport(UIKit)
import GameController

// TODO: Make a class for gamepad and another for joystick to support dpad and joystick as seperate inputs
@objc @objcMembers
public final class PViCadeGamepadDirectionPad: GCControllerDirectionPad {
    var handler: GCControllerDirectionPadValueChangedHandler?
    let _xAxis: PViCadeAxisInput = PViCadeAxisInput()
    let _yAxis: PViCadeAxisInput = PViCadeAxisInput()
    let _up: PViCadeGamepadButtonInput = PViCadeGamepadButtonInput()
    let _down: PViCadeGamepadButtonInput = PViCadeGamepadButtonInput()
    let _left: PViCadeGamepadButtonInput = PViCadeGamepadButtonInput()
    let _right: PViCadeGamepadButtonInput = PViCadeGamepadButtonInput()

    func padChanged(_ controllerIndex: Int = 0) {
        let state: iCadeControllerState = PViCadeReader.shared.states[controllerIndex]

        let x: Float = state.contains(.joystickLeft) ? -1.0 : state.contains(.joystickRight) ? 1.0 : 0.0
        let y: Float = state.contains(.joystickDown) ? -1.0 : state.contains(.joystickUp) ? 1.0 : 0.0

        _xAxis.value = x
        _yAxis.value = y

        _up.isPressed = y == 1
        _down.isPressed = y == -1
        _left.isPressed = x == -1
        _right.isPressed = x == 1

        handler?(self, x, y)
    }

    public override var xAxis: GCControllerAxisInput {
        return _xAxis
    }

    public override var yAxis: GCControllerAxisInput {
        return _yAxis
    }

    public override var up: GCControllerButtonInput {
        return _up
    }

    public override var down: GCControllerButtonInput {
        return _down
    }

    public override var left: GCControllerButtonInput {
        return _left
    }

    public override var right: GCControllerButtonInput {
        return _right
    }

    public override var valueChangedHandler: GCControllerDirectionPadValueChangedHandler? {
        get {
            return super.valueChangedHandler
        }
        set(handler) {
            self.handler = handler
        }
    }
}
#endif
